### DISCUSSION TITLE
Single molecule m6A probability predictions

### DISCUSSION THREAD

**Discussion by Stakaitis** [GoekeLab/m6anet#29] [Q&A] [answered]
Hi @chrishendra93,

Is there a script for single molecule m6A probability prediction? It would be very useful to know the percentage of modified/unmodified sites in a transcript.

Thanks,
Rytis Stakaitis

**Comment by chrishendra93** [comment]
hi @Stakaitis, apology for the late reply. I have yet to provide the script for this functionality but thinking of providing this in some format in the future.

Currently I'd say that the single molecule prediction is a bit noisy and not well calibrated (i.e the probability 0.5 does not reflect 50% of the times it is modified), and maybe the single molecule representation (the representation extracted from the last layer before pooling and classification in the network) might be more useful if you can perhaps do some clustering on it, then again I have yet to explore this

May I know what specific use cases you are interested to use this for so that I can take this into account for the next release?

Regards

Christopher Hendra

**Reply by Stakaitis** [reply]
Thank you for the reply,

I understand that the m6A modification calling in an individual read can have a lot of false positives/negatives, I think the same is with the [Epinano 1.2](https://github.com/novoalab/EpiNano). However, some sort of estimation on the proportion of modified/unmodified transcripts in a given sample, would be useful in treatment efficiency, disease progression, or tumor malignancy research.

In my particular case, I want to determine whether I can detect changes in m6A modification levels in specific transcripts during tumor progression. In other words, I want to see if transcript/gene X has more m6A modifications in high malignancy cancer vs low malignancy cancer.

If I understand correctly, now the final result from m6Anet shows a probability that a specific site of a transcript is modified. In this case, the information on m6A modification percentage is lost. Then, I assume, that in the scenario where there are 2 samples, and sample A has 5/100 m6A modifications at a site Y of a transcript X, and sample B has 35/100 m6A modifications at the same location, both of these samples will show the probability that site Y of a transcript X is unmodified. But these additional 30 modifications could be the cause of initiation of a certain biological processes - which could help detect early stages of various diseases, like cancer.

Rytis Stakaitis

**Comment by chrishendra93** [comment]
Hi @Stakaitis , sorry for the slow response, I was only back recently

I am planning to add this feature where you can output the individual read probability along with the site probability, hopefully in the near future (1 or 2 weeks time if everything goes well). I am thinking of outputting individual read probability and let the user decides the threshold to determine whether a read is modified or not (we will be providing a recommended value of course). Hopefully it will be useful

Regards

Christopher Hendra

**Comment by chrishendra93** [comment, marked as answer]
hi @Stakaitis, we have just released a new version of m6anet that can perform single molecule quantification. Do try it and let us know if it's working for you

Regards

Christopher Hendra

**Reply by Stakaitis** [reply]
Thank you! I will try to run it on my data soon and let you know if I notice any issues.

**Reply by Stakaitis** [reply]
Also, is this update also available on `nf-core/nanoseq -r dev` or is it more complicated to update that pipeline?

**Reply by chrishendra93** [reply]
hi @Stakaitis , the update is currently not available through nanoseq. You have to install this separately from github / pypi for now, but we're working to make the update available through nanoseq, hopefully within this month or next month

**Comment by MaestSi** [comment]
Hi @chrishendra93 and @Stakaitis ,
I just installed m6anet following installation instructions and ran m6anet-dataprep and m6anet-run_inference on the demo dataset. However, I am not able to find read-level modification results. Is there another script I should run? I was not able to find additional information in the docs.
Thanks in advance,
Simone

**Reply by Stakaitis** [reply]
Hi @MaestSi, 
I'm not a creator of m6Anet, I'm just a user, so my knowledge on this software is limited.
However, I was able to extract m6A modification values at a single-molecule level (it's probably the same as read-level?). I've done that in May-June of 2022, so something might have changed since then in the m6Anet software.
The steps I took to extract these modifications from dRNA-seq:

1. Base-call .fast5 files in High Accuracy (hac) mode with Guppy:
```
Config_file=rna_r9.4.1_70bps_hac.cfg
guppy_basecaller -i <dir_to_fast5_pass> -s <output_dir> -c ${Config_file} --fast5_out --num_callers 16 
```
2. Convert all .fastq files to .fasta using [common tools](https://bioinformaticsworkbook.org/dataWrangling/fastaq-manipulations/converting-fastq-format-to-fasta.html#gsc.tab=0)
3. Combine all .fasta files to one big .fasta file:
`cat /dir/to/fasta/*.fasta > new_id.fasta`
4. Map reads to full (ncRNA+cDNA) reference transcriptome with minimap2; index aligned files with samtools; index and eventalign with nanopolish:
```
ref=Homo_sapiens.GRCh38.cdna.ncrna.fa
fasta=new_id.fasta
bam=new_id.bam
minimap2 -ax splice -uf -k14 ${ref} ${fasta} | samtools sort -o ${bam}
samtools index ${bam}
nanopolish index --directory <dir_to_basecalled_fast5_files> ${fasta}
nanopolish eventalign --reads ${fasta} --bam ${bam} --genome ${ref} --scale-events --signal-index > <eventalign_output_file.txt>
```
5. Process data with m6Anet:
```
m6anet-dataprep --eventalign <eventalign_output_file.txt> --out_dir <dataprep_dir>
m6anet-run_inference --input_dir <dataprep_dir> --out_dir <inference_dir> **--infer_mod_rate**
gzip -d  <inference_dir>/data.result.csv.gz
```

Hope that helps

**Reply by MaestSi** [reply]
Dear @Stakaitis ,
I've already done what you kindly described. Indeed, in the data.result.csv file, there is a column reporting the modification ratio. However, I was looking for a read-to-mod-probability association, otherwise I can only infer the fraction of modified reads for each site but I can't extract mod/unmod reads.
Thanks,
SM

**Reply by chrishendra93** [reply]
hi @MaestSi, if you're looking for a feature that can label each read as modified, you can try the branch named read_id. It is not fully tested yet, mind you, but this will give you an additional file with the read index of each read (as per nanopolish read index, which if you have summary.txt from nanopolish eventalign you can map it back to the actual read ids) and the probability modified for each one. You might need to set a threshold to call the modification on the read level (our threshold is 0.033379376)

**Reply by MaestSi** [reply]
Thank you @chrishendra93 !
I git cloned the read_id branch and ran `python setup.py install`.
I then ran the code on demo data.
Although the m6anet-dataprep step ran smoothly, the m6anet-run_inference step produced the following error:
```
Traceback (most recent call last):
  File "/opt/conda/bin/m6anet-run_inference", line 33, in <module>
    sys.exit(load_entry_point('m6anet==1.1.1', 'console_scripts', 'm6anet-run_inference')())
  File "/opt/conda/bin/m6anet-run_inference", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/opt/conda/lib/python3.9/importlib/metadata.py", line 86, in load
    module = import_module(match.group('module'))
  File "/opt/conda/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "<frozen zipimport>", line 259, in load_module
  File "/opt/conda/lib/python3.9/site-packages/m6anet-1.1.1-py3.9.egg/m6anet/scripts/run_inference.py", line 2, in <module>
  File "/opt/conda/lib/python3.9/site-packages/torch-1.13.0-py3.9-linux-x86_64.egg/torch/__init__.py", line 191, in <module>
    _load_global_deps()
  File "/opt/conda/lib/python3.9/site-packages/torch-1.13.0-py3.9-linux-x86_64.egg/torch/__init__.py", line 153, in _load_global_deps
    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
  File "/opt/conda/lib/python3.9/ctypes/__init__.py", line 382, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libcublas.so.11: cannot open shared object file: No such file or directory
```
After installing libcublas with `conda install -c nvidia libcublas` I obtained similar error but with the last line:
```
OSError: libcudnn.so.8: cannot open shared object file: No such file or directory
```
This time, running `conda install -c conda-forge cudnn` did not solve the issue.
By the way, I do not have dedicated GPUs in my system.
Thanks,
Simone

**Reply by chrishendra93** [reply]
@MaestSi, ah yes, one way to change that is to replace the torch requirement in the setup.py file to 'torch==1.6.0'. It should fix the problem

**Reply by MaestSi** [reply]
Dear @chrishendra93 , the installation resulted in the following error:
```
#7 5.977 Searching for torch==1.6.0
#7 5.977 Reading https://pypi.org/simple/torch/
#7 7.828 No local packages or working download links found for torch==1.6.0
#7 7.829 error: Could not find suitable distribution for Requirement.parse('torch==1.6.0')
```
Is there another torch version which may be available and may work?
Thanks,
SM

**Reply by chrishendra93** [reply]
hi @MaestSi , I've just pushed a change to read_id branch to hopefully fix that problem, try pulling the latest version of the branch and run setup.py again

**Reply by MaestSi** [reply]
Hi, I got the same error, it could not find torch==1.6.0.
SM

**Reply by chrishendra93** [reply]
that is really weird since if you're running m6anet from the main branch, the torch version is 1.6.0. Maybe you can try to change it to other version, maybe 1.7.0? The problem that you encountered before was due to the setup installing the gpu version of the latest pytorch I think

**Reply by MaestSi** [reply]
Hi, I was able to get a CPU version of torch with:
`pip install http://download.pytorch.org/whl/cpu/torch-1.10.2%2Bcpu-cp39-cp39-linux_x86_64.whl`
Then, I modified the required torch version with:
`sed -i "s/torch==1.6.0/torch>=1.6.0/" setup.py`
However, command:
`python setup.py install`
results in compilation errors.

```
#9 63.53 error: Setup script exited with error: Command "ar rcs build/temp.linux-x86_64-3.9/libarpack_scipy.a build/temp.linux-x86_64-3.9/tmp/easy_install-ma420phk/scipy-1.4.1/scipy/_build_utils/src/wrap_g77_abi_c.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cgetv0.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cnaitr.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cnapps.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cnaup2.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cnaupd.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cneigh.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cneupd.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cngets.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/csortc.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cstatn.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dgetv0.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dnaitr.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dnapps.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dnaup2.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dnaupd.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dnconv.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dneigh.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dneupd.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dngets.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dsaitr.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dsapps.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dsaup2.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dsaupd.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dsconv.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dseigt.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dsesrt.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dseupd.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dsgets.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dsortc.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dsortr.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dstatn.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dstats.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dstqrb.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/sgetv0.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/snaitr.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/snapps.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/snaup2.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/snaupd.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/snconv.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/sneigh.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/sneupd.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/sngets.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/ssaitr.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/ssapps.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/ssaup2.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/ssaupd.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/ssconv.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/sseigt.o build/temp.linux-x86_64-3.9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/ssesrt.o" failed with exit status 1
#9 63.58 /opt/conda/lib/python3.9/site-packages/numpy/distutils/ccompiler_opt.py:647: UserWarning: The NumPy module was reloaded (imported a second time). This can in some cases result in small but subtle issues and is discouraged.
#9 63.58   from numpy.distutils import log
...
```
Do you have any ideas for solving this issue?
Thanks,
Simone

**Reply by chrishendra93** [reply]
hi @MaestSi, I can build the branch just fine by myself. I am just wondering if this might be due to the python version. Do you think you can try to reclone the repository and go to the branch read_id again and maybe run the setup.py using python 3.8?

**Reply by MaestSi** [reply]
Hi @chrishendra93 , with python 3.8 it works! I was able to obtain data.indiv_proba.csv.gz file, with read_index field, which I should be able to map back to the read id from summary.txt file generated by nanopolish eventalign. Thanks a lot for the support!
Simone

**Reply by chrishendra93** [reply]
hi @MaestSi, glad that it works for you. Let me know if you have any feedback on this, I am planning to include this feature in the next release

**Reply by WeipengMO** [reply]
> hi @MaestSi, if you're looking for a feature that can label each read as modified, you can try the branch named read_id. It is not fully tested yet, mind you, but this will give you an additional file with the read index of each read (as per nanopolish read index, which if you have summary.txt from nanopolish eventalign you can map it back to the actual read ids) and the probability modified for each one. You might need to set a threshold to call the modification on the read level (our threshold is 0.033379376)

Hi @chrishendra93 , thank you for developing such an excellent tool. I have tried the "read_id" branch. And got the "data.indiv_proba.csv.gz" results file. I want to know how to map the read index to the corresponding read_id?

This is the `eventalign.txt` file from `nanopolish eventalign`:

```
head eventalign.txt | column -t
contig       position  reference_kmer  read_index  strand  event_index  event_level_mean  event_stdv  event_length  model_kmer  model_mean  model_stdv  standardized_level  start_idx  end_idx
AT1G01020.2  81        GTGAA           7           t       1            92.95             4.060       0.00332       GTGAA       101.50      4.49        -1.79               48992      49002
AT1G01020.2  81        GTGAA           7           t       2            101.64            3.559       0.00664       GTGAA       101.50      4.49        0.03                48972      48992
AT1G01020.2  81        GTGAA           7           t       3            108.44            3.056       0.00299       GTGAA       101.50      4.49        1.46                48963      48972
AT1G01020.2  82        TGAAC           7           t       4            131.81            3.539       0.00398       TGAAC       118.66      5.56        2.23                48951      48963
AT1G01020.2  82        TGAAC           7           t       5            124.34            4.994       0.00232       TGAAC       118.66      5.56        0.96                48944      48951
AT1G01020.2  82        TGAAC           7           t       6            129.49            5.299       0.00531       TGAAC       118.66      5.56        1.83                48928      48944
AT1G01020.2  82        TGAAC           7           t       7            109.05            8.522       0.00232       TGAAC       118.66      5.56        -1.63               48921      48928
AT1G01020.2  82        TGAAC           7           t       8            126.58            8.079       0.01129       TGAAC       118.66      5.56        1.34                48887      48921
AT1G01020.2  83        GAACA           7           t       9            96.93             2.319       0.00531       GAACA       95.52       3.45        0.39                48871      48887
```

Thank you very much!

**Reply by WeipengMO** [reply]
I think I got the solution. Add `--summary summary.txt` parameter to `nanopolish eventalign` command.

**Reply by chrishendra93** [reply]
hi @WeipengMO, you are absolutely right. The summary.txt provides a mapping between read_id and read index, probably it is a good idea to provide this mapping in the next release. 

For the two of you that need this feature @WeipengMO and @MaestSi , can I just check what are you planning to use the individual read probability for? Just curious

**Reply by WeipengMO** [reply]
Thank you for your kind response @chrishendra93 . I'm trying some experimental test. I want to use the m6A methyltransferase to label RNA and check whether I can check this signal in single-molecule level.

**Reply by MaestSi** [reply]
Hi @chrishendra93 , no concrete plan at the moment yet, but it will be nice to find a biological application for this new exciting feature!

**Comment by MaestSi** [comment]
Dear @chrishendra93 ,
I just saw that read_id branch of m6anet was removed. Does the main branch provide single-read results now?
Thanks,
Simone

**Reply by chrishendra93** [reply]
hi @MaestSi 

Yeah I have just merged the read_id branch into the development branch. It will support single-read results by default now (no need additional flag) and the latest implementation also improves the speed for high --num_iterations options so if you want to filter the read modification by selecting sites with high probability_modified number, you can set higher num_iterations to make it more reliable

**Reply by MaestSi** [reply]
Thanks for the information!
SM
